### ISSUE TITLE
Autocorrect

### ISSUE THREAD

**Issue by siccovansas** [FossifyOrg/Keyboard#58]
### Checklist

- [X] I made sure that there are **no existing issues** - [open](https://github.com/FossifyOrg/Keyboard/issues) or [closed](https://github.com/FossifyOrg/Keyboard/issues?q=is%3Aissue+is%3Aclosed) - to which I could contribute my information.
- [X] I made sure that there are **no existing discussions** - [open](https://github.com/FossifyOrg/Keyboard/discussions) or [closed](https://github.com/FossifyOrg/Keyboard/discussions?discussions_q=is%3Aclosed) - to which I could contribute my information.
- [X] I have read the FAQs inside the app (Menu -> About -> FAQs) and my problem isn't listed.
- [X] **I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.**
- [X] This issue contains only one feature request.
- [X] I have read and understood the [contribution guidelines](https://github.com/FossifyOrg/Keyboard/blob/main/CONTRIBUTING.md).
- [X] I optionally [donated](https://fossify.org/donate) to support the Fossify mission.

### Feature description

It would be great to have autocorrect in the Keyboard app.

### Why do you want this feature?

Many popular keyboards have autocorrect, so people are used to that. I order for people to switch to Fossify Keyboard, many people probably want to have autocorrect.

**Comment by coderph0x** [comment]
I'd go for a suggestion of autocorrected words in the top bar. I absolutely hate it when unknown words get autocorrected in something absolutely unrelated as soon as I hit the spacebar.

**Comment by naveensingh** [comment]
If and when implemented, anything like autocorrect or a suggestion strip will always be optional.

**Comment by Ajt8000** [comment]
I second this. Auto correct is really needed in the keyboard.

**Comment by PanderMusubi** [comment]
Next to autosuggest, autocomplete and autocorrect based on a language specific dictionary, https://github.com/AnySoftKeyboard/AnySoftKeyboard also offers this for first and last names in your address book. Probably another feature request but easier to implement and start with as no language specific dictionary is needed.

**Comment by tswistak** [comment]
The challenge here would be how to implement dictionaries, since Fossify is offline by design. I've seen two approaches in FOSS keyboards that may be considered here:
- AnySoftKeyboard has separate APKs with dictionaries for other languages than English (English is built-in). You can see, there are lots of them on F-Droid: https://search.f-droid.org/?q=anysoftkeyboard&page=1&lang=pl. That approach would be more user-friendly, but I don't know how hard it would be to implement dictionaries as plugins.
- Heliboard (which is nearly like Fossify Keyboard but with dictionaries) links from the app to the page (https://codeberg.org/Helium314/aosp-dictionaries#dictionaries), where user can download a dictionary and then load it to the app. It's less user-friendly, but definitely it should be easy to implement.

**Comment by PanderMusubi** [comment]
If the dictionaries aren't that big for contemporary app size, just include them all in the app. Saves a lot of trouble.

**Comment by PanderMusubi** [comment]
E.g. Dutch word list is relatively large because concatenations of noun don't have a space in between them. Uncompressed word list of ~420,000 words is about 5 MB and compressed 1.4 MB.

**Comment by user4andre** [comment]
Please don't make the app unnecessarily bloated (more powerful than needed). In my opinion, simplicity is the strongest argument in favor of this keyboard. 

I use Anisoft Keyboard. The main reason I want to switch to Fossify is that there are no word suggestions. 
For me that is an advantage.

**Comment by acidrums4** [comment]
Taking in account that the very first sentence in the README of this project (and its web) it is stated that Fossify Keyboard is

> your go-to solution for effortless and efficient typing

then *this feature should be a must*, wether if enabled by default or made it optional in some setting whatsoever. But definitely a feature that in fact makes typing 'effortless and efficient'. I'd go as far as to say this should be top priority.

Now, there's a comment saying that this feature *should* be optional, and that of course is up to the developers, though it'be great if they make such option discoverable for every user and would-to-be user. But in that spirit I'd imagine Heliboard's approach would fit better - making the user download the dictionaries they'd use.

**Comment by HerrJM** [comment]
> If the dictionaries aren't that big for contemporary app size, just include them all in the app. Saves a lot of trouble.
> E.g. Dutch word list is relatively large because concatenations of noun don't have a space in between them. Uncompressed word list of ~420,000 words is about 5 MB and compressed 1.4 MB.

The easiest way would still be to choose which language(s) you want to download. E.g. I only need three, so why should I download all of them. It could be as simple as choosing the layout.

**Comment by PanderMusubi** [comment]
Note that keyboards work better with preprocessed word lists, not plain word lists. Also, autocorrection is an important part. For now, best is to use AnySoftKeyboard or look at the functionality/implementaiton there.

**Comment by ghost** [comment]
Are there any updates on this?

**Comment by naveensingh** [comment]
No, no updates yet.

**Comment by cig2000** [comment]
Hi, recently found Fossify apps and enjoy most of them but not having autocorrect option in the keyboard is causing me to uninstall it. Is app still being worked on?

**Comment by ghost** [comment]
Dude, it's open source. If this is so important to you, fork the repo and inplement it yourself. Otherwise be humble that people work on this software during their spare time and use it as is.

**Comment by cig2000** [comment]
> Dude, it's open source. If this is so important to you, fork the repo and inplement it yourself. Otherwise be humble that people work on this software during their spare time and use it as is.

It's been over a year since this feature request. I'm just curious if it's still being worked on or not.

**Comment by ghost** [comment]
Just read my last post again.

**Comment by naveensingh** [comment]
> Is app still being worked on?

Yes (see unreleased changes: https://github.com/FossifyOrg/Keyboard/pull/244)

> It's been over a year since this feature request. I'm just curious if it's still being worked on or not.

Recently, I have been thinking about how to implement this, given the [constraints](https://github.com/FossifyOrg/Keyboard/issues/58#issuecomment-2308449261). I will post when I have a solution.

**Comment by cig2000** [comment]
> > Is app still being worked on?
> 
> Yes (see unreleased changes: [#244](https://github.com/FossifyOrg/Keyboard/pull/244))
> 
> > It's been over a year since this feature request. I'm just curious if it's still being worked on or not.
> 
> Recently, I have been thinking about how to implement this, given the [constraints](https://github.com/FossifyOrg/Keyboard/issues/58#issuecomment-2308449261). I will post when I have a solution.

Thank you for the update! Great work!!

**Comment by gri38** [comment]
> Just read my last post again.
Hello @coderph0x ,

I think @cig2000  is perfectly entitled to ask questions about the “roadmap visibility” and to check whether the project is still active, in order to decide whether to follow it as a user—not necessarily as a developer.

Open source is fantastic, but contributions and questions in open source projects don’t come only from developers, and fortunately! Users can ask questions too.

I just wanted to “defend” @cig2000 because his tone was completely appropriate, and I felt your response might have been a bit harsh. I agree with you that communication is important, and indeed, it’s unacceptable when people complain about freely produced work. But in this case, I don’t think @cig2000 was doing that at all.

**Comment by ghost** [comment]
It's your right to look at it like that, I read the tone differently. Especially with his pressing "forcing me to uninstall it" and "it's been a year". That is entitled indeed, nothing less.

**Comment by calebbuening** [comment]
Love this project! @naveensingh let me know if there is anything I can help with.

**Comment by sahin52** [comment]
We can at least put a dictionary for English

**Comment by PanderMusubi** [comment]
If you do, please use the same format dictionary as https://github.com/AnySoftKeyboard/AnySoftKeyboard is using.

That will:
- save space (downloading and installed)
- make spell checking and autocompletion much faster than a flat list
- this keyboard can use directly a lot of existing dictionaries (and autoreplace additions and exceptions)

Those dictionaries can be used with proper attribution or when contacting the makers (such as myself being one) will contribute dictionaries to this app directly.

**Comment by sahin52** [comment]
Trying seconsm time givingnup again fprnthe same reason, there is no suggestionmstrip and itmis hard tp write eiyhoutmit, as you seemthis is written with this keybpars. Itmis necessaryntp havr a sugfwstion strip

**Comment by jasper-scelsi** [comment]
I think it makes sense to have the offline dictionary available, just say the download size and note because it is an offline app, a dictionary must be installed.
